dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,672 @@
+                // This should only happen if the value is irrelevant, so we will just pick 0.
+                case "false", "null" -> BigInteger.ZERO;
+                case "true" -> BigInteger.ONE;
+                default -> new BigInteger(valueString);
+            };
+
+            if (e.hasTag(Tag.READ)) {
+                em = new LoadModel(e, tm, id, address, value);
+                executionModel.addAddressRead(address, (LoadModel) em);
+            } else if (e.hasTag(Tag.WRITE)) {
+                em = new StoreModel(e, tm, id, address, value);
+                executionModel.addAddressWrite(address, (StoreModel) em);
+            } else {
+                // Should never happen.
+                logger.warn("Event {} has Tag MEMORY but no READ or WRITE", e);
+                em = new MemoryEventModel(e, tm, id, address, value);
+            }
+
+        } else if (e.hasTag(Tag.FENCE)) {
+            final String name = ((GenericVisibleEvent) e).getName();
+            em = new FenceModel(e, tm, id, name);
+        } else if (e instanceof Assert assrt) {
+            em = new AssertModel(assrt, tm, id);
+        } else if (e instanceof Local local) {
+            em = new LocalModel(local, tm, id);
+        } else if (e instanceof CondJump cj) {
+            em = new CondJumpModel(cj, tm, id);
+        } else {
+            // Should never happen.
+            logger.warn("Extracting the event {} that should not be extracted", e);
+            em = new DefaultEventModel(e, tm, id);
+        }
+
+        executionModel.addEvent(e, em);
+    }
+
+    private boolean toExtract(Event e) {
+        // We extract visible events, Locals and Asserts to show them in the witness,
+        // and extract also CondJumps for tracking internal dependencies.
+        return e.hasTag(Tag.VISIBLE)
+               || e instanceof Local
+               || e instanceof Assert
+               || e instanceof CondJump;
+    }
+
+    private void extractMemoryLayout() {
+        for (MemoryObject obj : context.getTask().getProgram().getMemory().getObjects()) {
+            boolean isAllocated = obj.isStaticallyAllocated()
+                                        || isTrue(context.execution(obj.getAllocationSite()));
+            if (isAllocated) {
+                BigInteger address = (BigInteger) evaluateByModel(context.address(obj));
+                BigInteger size = (BigInteger) evaluateByModel(context.size(obj));
+                executionModel.addMemoryObject(obj, new MemoryObjectModel(obj, address, size));
+            }
+        }
+    }
+
+    // Getting the correct relation to extract is tricky.
+    // In the case of redefinition, we have to find the latest defined one which we care about only.
+    // If there is no redefinition the original one will be returned simply.
+    private Relation getRelationWithName(String name) {
+        // First check if the original definition is asked.
+        if (name.endsWith("#0")) {
```

Comment:
It's not the same if there are redefinitions, because `X` will return the last one whereas `X#0` is always the first one. Whether the user actually uses this feature is a different question, but I think at least you and me are also users of the tool and being able to print intermediate relations seems very useable, especially when developing new models for GPUs.